### ISSUE TITLE
Extracting a 2048 dim feature vector from resnet50 

### ISSUE THREAD

**Issue by QuantScientist** [Tencent/ncnn#2283]
Hello :) 
I used the pre-trained pytorch resnet50 model, converted to onnx and then to ncnn using the online converter at https://convertmodel.com/

I am able to load the model, but whatever I do I  can not access this convolution. 

<img width="668" alt="2020-11-04 14_34_03-resnet50-opt param" src="https://user-images.githubusercontent.com/18743986/98112924-91d79e00-1eab-11eb-95e3-3576bcd42959.png">

The resulting param file clearly shows the name as:
`Convolution              Conv_116                 1 1 488 652 0=2048 1=1 5=1 6=1048576`

```
Convolution              Conv_114                 1 1 485 488 0=512 1=3 4=1 5=1 6=2359296 9=1
Convolution              Conv_116                 1 1 488 652 0=2048 1=1 5=1 6=1048576
BinaryOp                 Add_117                  2 1 652 482_splitncnn_0 491
ReLU                     Relu_118                 1 1 491 492
Pooling                  GlobalAveragePool_119    1 1 492 494 0=1 4=1
InnerProduct             Gemm_121                 1 1 494 495 0=1000 1=1 2=2048000 
```

The code I am using is:
```
ncnn::Extractor ex = net.create_extractor();
	ex.input("input.1", in); // this works, why? 
	ncnn::Mat out;
//    for ( auto i = net.blobs.begin(); i != net.blobs.end(); i++ )
//    {
//        std::cout << i->name << std::endl;
//        std::cout << i->producer << std::endl;
//    }
	ex.extract("Conv_116", out); // Convolution Conv_116 1 1 488 652 0=2048 1=1 5=1

I also tried:
ex.extract("488", out); 
TraceLog(LOG_INFO, "ncnnRay: fv2048 out =%d", out.w); 
std::vector<float> vec;
for (int i = 0; i < out.w; ++i) {
       vec.emplace_back(*(static_cast<float*>(out.data) + i));
 }
TraceLog(LOG_INFO, "ncnnRay: fv2048 vec =%d", vec.size());
```

But I am getting 16 and not geting a vector of 2048. 
Thanks,

**Comment by nihui** [comment]
```Conv_116``` is the layer name
You need to pass the output blob name to ex.extract(), in this case, it shall be ```652```, which is shown in the outputs section in netron

**Comment by QuantScientist** [comment]
Got it, thanks. 
Is there a fatser way to copy from out to a double vector/ array, I mean like memcpy? 
```
ex.extract("652", out); // Convolution Conv_116 1 1 488 652 0=2048 1=1 5=1
	TraceLog(LOG_INFO, "ncnnRay: fv2048 out.c =%d", out.c);
    std::vector<float> vec;
    for (int i = 0; i < out.c; ++i) {
        vec.emplace_back(*(static_cast<float*>(out.data) + i));
    }
    TraceLog(LOG_INFO, "ncnnRay: fv2048 vec =%d", vec.size());
```

**Comment by QuantScientist** [comment]
Well,
Still not working, all elements are zero inside the 2048 vector. 

What I really wanted to do is to extract a feature vector like so:
```
import torch
import torchvision
import torch.onnx

class ResNetBottom(torch.nn.Module):
    def __init__(self, original_model, real_name, num_feats):
        super(ResNetBottom, self).__init__()
        self.features = torch.nn.Sequential(*list(original_model.children())[:-1])
        self.real_name = real_name
        self.num_feats = num_feats
    def forward(self, x):
        x = self.features(x)
        x = x.view(x.size(0), -1)
        return x

res_model = torchvision.models.resnet50(pretrained=True)
res_conv2 = ResNetBottom(res_model, real_name="res52", num_feats=2048)
for param in res_conv2.parameters():
    param.requires_grad = False
res_conv2.eval()
model=res_conv2
x = torch.rand(1, 3, 224, 224)
torch_out = torch.onnx._export(model, x, "resnet50.onnx", export_params=True)
```

But when I convert it to ncnn I get: 

```
Shape not supported yet!Gather not supported yet! # axis=101
Unsupported unsqueeze axes !
```

**Comment by QuantScientist** [comment]
This is the ResNetBottom NN. It can not be converted to ncnn. 
[
<img width="651" alt="2020-11-04 18_08_15-resnet50 onnx" src="https://user-images.githubusercontent.com/18743986/98135709-d291e000-1ec8-11eb-8bd1-5ba65beda57d.png">
](url)

This is the onnx file:
https://github.com/QuantScientist/ncnnRay/blob/main/python_to_cpp/resnet50.onnx

Can you help?

**Comment by nihui** [comment]
use onnx-simplifier to eliminate these unsupported ops

https://github.com/Tencent/ncnn/wiki/use-ncnn-with-pytorch-or-onnx

**Comment by nihui** [comment]
> Got it, thanks.
> Is there a fatser way to copy from out to a double vector/ array, I mean like memcpy?
> 
> ```
> ex.extract("652", out); // Convolution Conv_116 1 1 488 652 0=2048 1=1 5=1
> 	TraceLog(LOG_INFO, "ncnnRay: fv2048 out.c =%d", out.c);
>     std::vector<float> vec;
>     for (int i = 0; i < out.c; ++i) {
>         vec.emplace_back(*(static_cast<float*>(out.data) + i));
>     }
>     TraceLog(LOG_INFO, "ncnnRay: fv2048 vec =%d", vec.size());
> ```

```
std::vector<float> vec(2048);
out = out.reshape(2048);
memcpy(vec.data(), out, 2048*sizeof(float));
```

**Comment by QuantScientist** [comment]
> use onnx-simplifier to eliminate these unsupported ops
> 
> https://github.com/Tencent/ncnn/wiki/use-ncnn-with-pytorch-or-onnx

Thanks, the simplifier at https://convertmodel.com/ is the one returning these errors when I upload https://github.com/QuantScientist/ncnnRay/blob/main/python_to_cpp/resnet50.onnx

**Comment by nihui** [comment]
Try to install and run onnx-simplifier locally.
The one in convertmodel.com is not up-to-date

**Comment by QuantScientist** [comment]
Thanks, 
All the code is here if you wana have a look:
https://github.com/QuantScientist/ncnnRay/tree/main/python_to_cpp
<img width="800" alt="2020-11-11 12_58_27-ncnnRay_ Model Studio" src="https://user-images.githubusercontent.com/18743986/98803733-e0da8180-241d-11eb-95c0-2a5a63b371f8.png">

**Comment by nihui** [comment]
针对onnx模型转换的各种问题，推荐使用最新的pnnx工具转换到ncnn
In view of various problems in onnx model conversion, it is recommended to use the latest pnnx tool to convert your model to ncnn

```shell
pip install pnnx
pnnx model.onnx inputshape=[1,3,224,224]
```

详细参考文档
Detailed reference documentation
https://github.com/pnnx/pnnx
https://github.com/Tencent/ncnn/wiki/use-ncnn-with-pytorch-or-onnx#how-to-use-pnnx
